ykrt/src/compile/jitc_yk/jit_ir.rs

```diff
@@ -26,297 +17,401 @@ const SHORT_INSTR_MAX_OPERANDS: u64 = 3;
 ///  - `*_SHIFT`: the number of bits required to left shift a field's value into position (from the
 ///  LSB).
 ///
-/// Bit fiddling for a short operands:
-const SHORT_OPERAND_SIZE: u64 = 18;
-const SHORT_OPERAND_KIND_SIZE: u64 = 3;
-const SHORT_OPERAND_KIND_MASK: u64 = 7;
-const SHORT_OPERAND_VALUE_SIZE: u64 = 15;
-const SHORT_OPERAND_VALUE_SHIFT: u64 = SHORT_OPERAND_KIND_SIZE;
-const SHORT_OPERAND_MAX_VALUE: u64 = !(u64::MAX << SHORT_OPERAND_VALUE_SIZE);
-const SHORT_OPERAND_MASK: u64 = 0x3ffff;
-/// Bit fiddling for instructions.
-const INSTR_ISSHORT_SIZE: u64 = 1;
-const INSTR_ISSHORT_MASK: u64 = 1;
-/// Bit fiddling for short instructions.
-const SHORT_INSTR_OPCODE_MASK: u64 = 0xe;
-const SHORT_INSTR_OPCODE_SHIFT: u64 = INSTR_ISSHORT_SIZE;
-const SHORT_INSTR_FIRST_OPERAND_SHIFT: u64 = INSTR_ISSHORT_SIZE + OPCODE_SIZE;
-
-/// An instruction is identified by its index in the instruction vector.
-#[derive(Copy, Clone)]
-pub(crate) struct InstructionID(usize);
-
-impl InstructionID {
-    pub(crate) fn new(v: usize) -> Self {
-        Self(v)
-    }
+const OPERAND_INDEX_MASK: u16 = 0x7fff;
+
+// The maximum size of an operand we can encode in an instruction is 15 bits.
+const MAX_OPERAND_IDX: u16 = 32768;
```

Comment:
It's impossible for someone stupid like to me to convert from decimal to a bit pattern. I personally normally prefer `0b111111` notation for this (unless there are "too many" bits for someone to count accurately) but hex `0x1234` notation or `x << y` shifting can work too. [Note that `OPERAND_INDEX_MASK` uses hex already!]

Reply:
It's the maximum value an operand index can assume. I don't think it makes sense to store a conceptually integer decimal value as a bit pattern or hex.

It makes sense for masks to be hex though.

Reply:
Right so `x << y` might be best here as it makes clearer that this isn't a random value but is a clear power of 2-1? In particular, I think the max value here is 1 greater than it should be.

Reply:
> I think the max value here is 1 greater than it should be.

Yes, I just realised that this is bogus. It should be all ones in the least-significant bits! Will fix.

Reply:
Fixed in c62a43fe486af2b46a2cff5828622d89de73ea06